From 7e9a93c7cc0b32d97fadf11f039ac0c46fb4ac84 Mon Sep 17 00:00:00 2001
From: "jian.xu" <jian.xu@amlogic.com>
Date: Sat, 12 Apr 2014 10:14:53 +0800
Subject: [PATCH 3899/5965] PD #90003: 1) fix hdmitx driver audio parameter
 async setting issue;2) fix raw data channel status setting issue

---
 drivers/amlogic/hdmi/hdmi_tx/hdmi_tx.c | 34 ++++++++++++++------------
 sound/soc/aml/m8/aml_spdif_dai.c       |  2 ++
 2 files changed, 21 insertions(+), 15 deletions(-)

diff --git a/drivers/amlogic/hdmi/hdmi_tx/hdmi_tx.c b/drivers/amlogic/hdmi/hdmi_tx/hdmi_tx.c
index d4a90296ba51..f72994462e6e 100755
--- a/drivers/amlogic/hdmi/hdmi_tx/hdmi_tx.c
+++ b/drivers/amlogic/hdmi/hdmi_tx/hdmi_tx.c
@@ -1041,6 +1041,24 @@ static int hdmitx_notify_callback_a(struct notifier_block *block, unsigned long
         hdmi_print(INF, AUD "no update\n");
     else
         hdmitx_device.audio_notify_flag = 1;
+
+
+    if((!hdmi_audio_off_flag)&&(hdmitx_device.audio_param_update_flag)) {
+        if(hdmitx_device.hpd_state == 1) {     // plug-in & update audio param
+            hdmitx_set_audio(&hdmitx_device, &(hdmitx_device.cur_audio_param), hdmi_ch);
+	    if((hdmitx_device.audio_notify_flag == 1) || (hdmitx_device.audio_step == 1)) {
+                hdmitx_device.audio_notify_flag = 0;
+                hdmitx_device.audio_step = 0;
+#ifndef CONFIG_AML_HDMI_TX_HDCP
+                hdmitx_device.HWOp.CntlConfig(&hdmitx_device, CONF_AUDIO_MUTE_OP, AUDIO_UNMUTE);
+#endif
+            }
+            hdmitx_device.audio_param_update_flag = 0;
+            hdmi_print(INF, AUD "set audio param\n");
+        }
+    }
+
+
     return 0;
 }
 
@@ -1099,21 +1117,6 @@ static int hdmi_task_handle(void *data)
             }
         }
 
-        if((!hdmi_audio_off_flag)&&(hdmitx_device->audio_param_update_flag)) {
-            if(hdmitx_device->hpd_state == 1) {     // plug-in & update audio param
-                hdmitx_set_audio(hdmitx_device, &(hdmitx_device->cur_audio_param), hdmi_ch);
-                if((hdmitx_device->audio_notify_flag == 1) || (hdmitx_device->audio_step == 1)) {
-                    hdmitx_device->audio_notify_flag = 0;
-                    hdmitx_device->audio_step = 0;
-#ifndef CONFIG_AML_HDMI_TX_HDCP
-                    hdmitx_device->HWOp.CntlConfig(hdmitx_device, CONF_AUDIO_MUTE_OP, AUDIO_UNMUTE);
-#endif
-                }
-                hdmitx_device->audio_param_update_flag = 0;
-                hdmi_print(INF, AUD "set audio param\n");
-            }
-        }
-
         if(hdmitx_device->hpd_state == 0) {
             hdmitx_device->HWOp.CntlMisc(hdmitx_device, MISC_TMDS_PHY_OP, TMDS_PHY_DISABLE);
         }
@@ -1167,6 +1170,7 @@ edid_op:
                 hdmitx_device->tv_no_edid = 0;
             }
             set_disp_mode_auto();
+            hdmitx_set_audio(hdmitx_device, &(hdmitx_device->cur_audio_param), hdmi_ch);
             switch_set_state(&sdev, 1);
             cec_node_init(hdmitx_device);
         }
diff --git a/sound/soc/aml/m8/aml_spdif_dai.c b/sound/soc/aml/m8/aml_spdif_dai.c
index ff4121e80aa3..e8f030d256bb 100755
--- a/sound/soc/aml/m8/aml_spdif_dai.c
+++ b/sound/soc/aml/m8/aml_spdif_dai.c
@@ -210,6 +210,8 @@ static void aml_hw_iec958_init(struct snd_pcm_substream *substream)
 		//audio_set_i2s_mode(AIU_I2S_MODE_PCM16);
 		//audio_set_aiubuf(start, size);		
 	}else{
+		set.chan_stat->chstat0_l = 0x1902;
+                set.chan_stat->chstat0_r = 0x1902;
 		if(IEC958_mode_codec == 4){ //DD+
 			if(runtime->rate == 32000){
 				set.chan_stat->chstat1_l = 0x300;
-- 
2.19.0

